API refactor + update

skull132 opened the pull request:
* Refactors the API to be more spread out over files, for easier navigation and searching.
* Makes the **get_serverstatus**, **ping**, and **get_count_player** commands public.
* Removes the player and admin lists from the get_serverstatus command, since it duplicates the get_player_list command.
* Removes the `in world` loop from the ping command.
* Adds an API interface to interact with tickets over. This includes refactoring a few ticket functions to adhere to DRY.
* Implements the rate-limiting fix from #5049 by using `REALTIMEOFDAY`.
* Refactored `api_do_auth_check` into the `topic_command/check_auth(addr, auth_key)`, so that it could be reused in the `api_explain_command` command.

Is meant to supersede #5049

skull132 commented:
!review

Karolis2011 commented on code/modules/world_api/helpers.dm:
```diff
@@ -0,0 +1,80 @@
+		log_debug("API: Throttling bypassed - Command [command.name] set to no_throttle")
+	else
+		if(world_api_rate_limit[addr] != null && config.api_rate_limit_whitelist[addr] == null) //Check if the ip is in the rate limiting list and not in the whitelist
+			if(abs(world_api_rate_limit[addr] - world.time) < config.api_rate_limit) //Check the last request time of the ip
```
Should implement rate limiting fix, if there is no players world.time stops.

skull132 replied:
Righto. Forgot about that.

Arrow768 commented on code/datums/api.dm:
```diff
@@ -134,1093 +134,6 @@ proc/api_update_command_database()
-	return 1
-
 //
 // API for the other stuff
```
Rouge Comment

Arrow768 commented on code/modules/world_api/commands/admin.dm:
```diff
@@ -0,0 +1,116 @@
+	C.discord_admin = queryparams["senderkey"]
+
+	C << 'sound/effects/adminhelp.ogg'
+	C << message
```
```suggestion
	to_chat(C,message)
```

Arrow768 commented on code/modules/world_api/commands/admin.dm:
```diff
@@ -0,0 +1,116 @@
+	C.received_discord_pm = world.time
+	C.discord_admin = queryparams["senderkey"]
+
+	C << 'sound/effects/adminhelp.ogg'
```
```suggestion
	sound_to(C,'sound/effects/adminhelp.ogg')
```

Arrow768 commented on code/modules/world_api/commands/server_query.dm:
```diff
@@ -0,0 +1,296 @@
+				positions["misc"] = list()
+			positions["misc"][name] = rank
+
+	// for(var/k in positions)
```
Dead Code

Arrow768 commented on code/modules/world_api/helpers.dm:
```diff
@@ -0,0 +1,80 @@
+	return 1 // Bad Key
+
+
+/proc/api_update_command_database()
```
You should probs move that to /datum/topic_command/update_command_database since it isnt used anywhere else

skull132 commented:
Updates:
* Removed the unused api.dm file under datums/.
* Addressed all other comments.
* Refactored `api_do_auth_check` into the `topic_command/check_auth(addr, auth_key)`, so that it could be reused in the `api_explain_command` command.

Arrow768 commented on code/modules/world_api/commands/api_helpers.dm:
```diff
@@ -131,11 +108,32 @@
-	api_update_command_database()
+	if (!api_update_command_database())
+		statuscode = 500
+		return TRUE
```
return false on error

Karolis2011 commented on code/modules/world_api/commands/server_query.dm:
```diff
@@ -0,0 +1,295 @@
+	s["gameid"] = game_id
+	s["game_state"] = SSticker ? 0 : SSticker.current_state
+	s["transferring"] = !emergency_shuttle_shuttle.online()
+
```
Could we have map and server revision here as available variable?

Karolis2011 commented on code/modules/world_api/commands/server_query.dm:
```diff
@@ -0,0 +1,295 @@
+	s["host"] = host ? host : null
+	s["stationtime"] = worldtime2text()
+	s["roundduration"] = get_round_duration_formatted()
+	s["gameid"] = game_id
```
Maybe as we are refactoring could be have a way to alter data depending if user is authed or not, like game ID is somewhat sensitive, no? Then we could add if that checks for Auth and not send game ID to non authed requests. This thing could be used to protect other somewhat sensitive data, like player lists, while maintaining compatibility.

skull132 replied:
None of this information is sensitive. And in theory, you can already add the sensitive data check by using a proc to check if the query is authed.

However, since end points should roughly do one thing, it might be a bad idea to split the information in two without setting up another endpoint.
